Clarify optionality of type annotations for writing a plugin (#367)
napari/docs: docs/plugins/building_a_plugin/_layer_data_guide.md

deconstructed into a {class}`tuple` that we refer to as a `LayerData` tuple.  This type shows
up often in plugins and is explained here.

> Note that when writing your own plugin, type annotations are nearly always optional, 
> except in rare cases like [`magicgui` function widgets](magicgui).

Reviewer: ```suggestion
Note that when writing your own plugin, type annotations are optional, 
except in the case of [`magicgui` function widgets](magicgui).
```

to be explicit in that there is only one case.

> Here we sometimes provide a type annotation with a more informative name, 
> indicating its value's _functional role_ in a plugin, in addition to its possible types.

Reviewer: You were right in that they are type aliases! Sorry :facepalm: 
So I think all npe2 type aliases are defined [here](https://github.com/napari/npe2/blob/main/src/npe2/types.py).

So maybe something like:
```suggestion
Napari defines several type aliases for `LayerData`  related types. These
provide more  names that indicates the value's _functional role_ in a
plugin. We describe them below.
```

technically its npe2 not napari but I think this is okay.

Author: I use a wording very similar to what you suggest, here: https://github.com/napari/docs/pull/367/commits/a8a053eb99b6e85c263611f09f94eb710546ca8f

> Here in this section we sometimes provide a type annotation with a more informative name, 
> indicating its value's _functional role_ in a plugin, in addition its possible types.

Reviewer: ```suggestion
Here we sometimes provide a type annotation with a more informative name, 
indicating its value's _functional role_ in a plugin, in addition to its possible types.
```
nitpick

deconstructed into a {class}`tuple` that we refer to as a `LayerData` tuple.  This type shows
up often in plugins and is explained here.

Note that when writing your own plugin, type annotations are optional, 
except in the case of [`magicgui` function widgets](magicgui).

> For several types related to `LayerData` tuples, Napari defines an alias 

Reviewer: ```suggestion
For several types related to `LayerData` tuples, Napari defines a type alias 
```

> For several types related to `LayerData` tuples, Napari defines a type alias 

Reviewer 2: ```suggestion
For several types related to `LayerData` tuples, napari defines a type alias 
```

napari is lower case!

where ...

```python
from typing import Dict, Literal, Protocol, Sequence, Tuple, Union
import numpy as np

LayerTypeName = Literal[
    "image", "labels", "points", "shapes", "surface", "tracks", "vectors"
]
LayerProps = Dict

> DataType = Union["ArrayLike", Sequence["ArrayLike"]]

Reviewer: Actually, let's leave these as is, as I don't think people will be copying this code snippet. This code is just to inform people of what the type aliases are.

Author: OK, changed here: https://github.com/napari/docs/pull/367/commits/ae13166ea76a93ae202cecb42ca34c90823cb9bc

# typing than literal array types (e.g. numpy, dask, xarray, etc...)
```

Note that when writing your own plugin, such type annotations are nearly always optional, 
except in rare cases like `magicgui` function widgets.

> Here, for clarity, we've provided aliases that indicate the how values of the type are 

Reviewer 2: Something is off with this line, can you take another look?

Author: yes sorry 😅 , ﻿﻿﻿[ec41dda66f3c8a55f647dbe5c07f852e26904cb2](https://github.com/napari/docs/pull/367/commits/ec41dda66f3c8a55f647dbe5c07f852e26904cb2)

> being used and the level of generality that's possible (e.g., with `ArrayLike`). 

Reviewer: Sorry I don't quite follow, what is meant by aliases here? Thanks

Author: Sorry, I just meant different names (with clearer context meaning) for the types. I changed here: 5ce242db122766b936b151ef328e7c65b7d19d5d
WDYT?

> being used and the level of generality that's possible (e.g., with `ArrayLike`). 

Reviewer: Thanks. What do you mean "how values of the type are being used"? As in the definition of each data type and their relationship with each other?

Author: I mean to refer to the meaning of the values of these types in the context of a napari plugin, rather than just generally. Like I think it's quite helpful that `Dict` is renamed as `LayerProps` to show the role that data in the dict are playing when used in the plugin, and that `FullDataLayer` relabels an otherwise relatively complex tuple type with a name that tells readers / plugin developers clearly how those data fit in the functioning of the plugin.

Maybe something like "the role of the corresponding values in a plugin" replacing "how values of the type are being used" would make this snippet more readable?

Reviewer: Ohhh I see what you are saying. You're right. I like your suggested wording!

It was not immediately clear to me that `LayerProps` stood for 'layer properties' actually.
